api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirImmunizationServiceImpl.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.api.impl;
+
+import static org.openmrs.module.fhir2.FhirConstants.CODED_SEARCH_HANDLER;
+import static org.openmrs.module.fhir2.FhirConstants.PATIENT_REFERENCE_SEARCH_HANDLER;
+import static org.openmrs.module.fhir2.api.translators.impl.ImmunizationTranslatorImpl.immunizationGroupingConcept;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import ca.uhn.fhir.rest.api.SortSpec;
+import ca.uhn.fhir.rest.param.ReferenceAndListParam;
+import ca.uhn.fhir.rest.param.TokenAndListParam;
+import ca.uhn.fhir.rest.param.TokenParam;
+import org.hl7.fhir.r4.model.Immunization;
+import org.openmrs.Concept;
+import org.openmrs.Obs;
+import org.openmrs.api.EncounterService;
+import org.openmrs.api.ObsService;
+import org.openmrs.module.fhir2.api.FhirImmunizationService;
+import org.openmrs.module.fhir2.api.dao.FhirObservationDao;
+import org.openmrs.module.fhir2.api.search.param.SearchParameterMap;
+import org.openmrs.module.fhir2.api.translators.ImmunizationTranslator;
+import org.openmrs.module.fhir2.api.translators.impl.ImmunizationObsGroupHelper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FhirImmunizationServiceImpl implements FhirImmunizationService {
+	
+	@Autowired
+	private ImmunizationTranslator translator;
+	
+	@Autowired
+	private FhirObservationDao obsDao;
+	
+	@Autowired
+	private ObsService obsService;
+	
+	@Autowired
+	private EncounterService encounterService;
+	
+	@Autowired
+	private ImmunizationObsGroupHelper helper;
+	
+	@Override
+	public Immunization get(String uuid) {
+		return translator.toFhirResource(obsDao.get(uuid));
+	}
+	
+	@Override
+	public Immunization create(Immunization newImmunization) {
+		Obs obs = translator.toOpenmrsType(newImmunization);
+		if (obs.getEncounter().getId() == null) {
+			encounterService.saveEncounter(obs.getEncounter());
+		}
+		obs = obsDao.createOrUpdate(obs);
+		return translator.toFhirResource(obs);
+	}
+	
+	@Override
+	public Immunization update(String uuid, Immunization updatedImmunization) {
+		Obs obs = translator.toOpenmrsType(obsDao.get(uuid), updatedImmunization);
+		obs = obsService.saveObs(obs, "Updated when translating a FHIR Immunization resource.");
+		//		obs = obsDao.createOrUpdate(obs);
```

Comment:
↑ above is the last reference to `ObsService`, however uncommenting the line that switches to using `FhirObservationDao` instead does throw the infamous "cannot change immutable obs" exception. Any idea?

Reply:
Hmmmm... yeah. This is actually why I'm not planning on implementing the "update" interaction for observations, mostly because the immutability requirement means that the UUID will change and so you have a PUT request operating to create a new resource and no good way to communicate that to the client. So, actually, my first recommendation would just be to not implement updates at all.

If we really do need to support updates to immunization resources, the only other alternative I could think of would be to essentially leave the main Obs itself alone, but void and replace the existing group members.

Reply:
@mks-d Apart from this unresolved issue, I think the code looks good.